morebits.js

```diff
@@ -4566,215 +4577,1725 @@ Morebits.wiki.page = function(pageName, status) {
+	};
+
+
+	/**
+	 * Construct the common base for block, unblock, and userrights
+	 * actions.  Includes an api post to watch a user for unblock and
+	 * userrights actions, as they do not support the watchuser option.
+	 *
+	 * @param {string} action - The action being undertaken (`block`, `unblock`, or `userrights`).
+	 * @returns {object} Action-specific POST query.
+	 */
+	var fnBaseAction = function(action) {
+		var query = {
+			action: action,
+			user: ctx.userName,
+			reason: ctx.reason,
+			token: action === 'userrights' ? ctx.userrightsToken : ctx.csrfToken,
+			format: 'json'
+		};
+		if (ctx.changeTags) {
+			query.tags = ctx.changeTags;
+		}
+		// block or userrights
+		if (action !== 'unblock' && (ctx.expiry || (Array.isArray(ctx.expiry) && ctx.expiry.length))) {
+			query.expiry = ctx.expiry;
+		}
+		if (ctx.watchuser) {
+			if (action === 'block') {
+				query.watchuser = ctx.watchuser;
+				if (ctx.watchlistExpiry) {
+					query.watchlistexpiry = ctx.watchlistExpiry;
+				}
+			} else {
+				// Dumb hack: watchlist options not supported for
+				// unblock [[phab:T257662]] or userrights [[phab:T272294]], so fake it.
+				var watch_query = {
+					action: 'watch',
+					titles: mw.Title.newFromText(ctx.userName, 2).toText(),
+					token: mw.user.tokens.get('watchToken')
+				};
+				if (ctx.watchlistExpiry) {
+					watch_query.expiry = ctx.watchlistExpiry;
+				}
+				new Morebits.wiki.api('Watching user page', watch_query).post();
+			}
+		}
+
+		return query;
+	};
+
+	/**
+	 * Determine whether we need to first load the user.  The only
+	 * exception is notifications that don't care whether the target user
+	 * is a bot or indefinitely blocked, or if the talk page if opted-out.
+	 *
+	 * @param {string} action - The action being undertaken, e.g. `notify`
+	 * or `block`.  Only `notify` has any meaning.
+	 * @returns {boolean}
+	 */
+	var fnDontNeedLoad = function(action) {
+		if (ctx.userLoaded ||
+			(action === 'notify' && ctx.notifyBots && ctx.notifyIndef && !ctx.notifySkipLink && (!ctx.notifySkipTemplates || ctx.notifySkipTemplates.length === 0))) {
+			return true;
+		}
+		return false;
+	};
+
+
+	/*
+	  Wrappers for fnSuccess, the joint success function.  At the moment,
+	  we're not doing anything unique for any of these, so this is just
+	  for the structure.  If we do want to customize for specific
+	  scenarios, they should be broken out.
+	*/
+	var fnBlockSuccess = function() {
+		fnSuccess('block');
+	};
+	var fnUnblockSuccess = function() {
+		fnSuccess('unblock');
+	};
+	var fnGroupsSuccess = function() {
+		fnSuccess('groups');
+	};
+	var fnSuccess = function(action) {
+		ctx.actionResponse = ctx[action + 'Api'].response;
+
+		// `block: block` and `unblock: unblock`, but `groups: userrights`
+		var exactName = action === 'groups' ? 'userrights' : action;
+		// The API thinks we're successful if there's a response for the action,
+		// i.e. there isn't `result: 'Success'` like action=edit
```

Comment:
Not necessarily relevant here, but the variability in successful API response among various actions is *maddening*, even down to different capitalization among parameters.